### ISSUE TITLE
Upgrade to Scala 2.13

### ISSUE THREAD

**Issue by zchbndcc9** [krzemin/octopus#27]
It would be nice to upgrade the Scala versioning so that it can be used with 2.13. What needs to be done in order to accomplish this?

**Comment by krzemin** [comment]
Well, first we need to be sure that all dependencies published versions for Scala 2.13. Then we must cross compile against updated dependencies. Finally, I can release artifacts for 2.13 as well.

As I'm short on time recently, PRs are welcome.

**Comment by joules-o** [comment]
@eltherion I would very much like to use octopus in a Scala 2.13 project, have you made any progress in figuring out why the build is failing?

**Comment by eltherion** [comment]
@joules-o Not yet, unfortunately. :( I wasn't checking the issue since then. Could you, please, make a simple test for me? I need another machine to perform it, but I don't have one.
Since the Scala 2.13.1 is available, please, run these commands:
```bash
sbt -mem 4096 ++2.13.1 clean test:compile
[info] Loading settings for project global-plugins from scala-steward.sbt ...
[info] Loading global plugins from /Users/user/.sbt/1.0/plugins
[info] Loading settings for project octopus-build from plugins.sbt ...
[info] Loading project definition from /Users/user/Projects/octopus/project
[info] Loading settings for project root from version.sbt,build.sbt ...
[info] Set current project to root (in build file:/Users/user/Projects/octopus/)
[info] Setting Scala version to 2.13.1 on 7 projects.
[info] Reapplying settings...
[info] Set current project to root (in build file:/Users/user/Projects/octopus/)
[success] Total time: 0 s, completed Sep 28, 2019 10:57:01 AM
[info] Compiling 13 Scala sources to /Users/user/Projects/octopus/octopus/.jvm/target/scala-2.13/classes ...
[info] Compiling 13 Scala sources to /Users/user/Projects/octopus/octopus/.js/target/scala-2.13/classes ...
[info] Compiling 2 Scala sources to /Users/user/Projects/octopus/octopusCats/.jvm/target/scala-2.13/classes ...
[info] Compiling 13 Scala sources to /Users/user/Projects/octopus/octopus/.jvm/target/scala-2.13/test-classes ...
[info] Compiling 2 Scala sources to /Users/user/Projects/octopus/octopusScalaz/.jvm/target/scala-2.13/classes ...
[info] Compiling 2 Scala sources to /Users/user/Projects/octopus/octopusCats/.js/target/scala-2.13/classes ...
[info] Compiling 13 Scala sources to /Users/user/Projects/octopus/octopus/.js/target/scala-2.13/test-classes ...
[info] Compiling 2 Scala sources to /Users/user/Projects/octopus/octopusScalaz/.js/target/scala-2.13/classes ...

  | => octopusJS / Test / compileIncremental 100s
  | => octopusJVM / Test / compileIncremental 100s
```
On my machine it takes ages and compiling test classes never ends. Is this reproducible in your case as well?

**Comment by joules-o** [comment]
@eltherion, that's what I found in my case too. I found one more instance of a deprecated `.right` on line 62 of `ValidatorSyntaxSpec` and hoped removing that would fix the problem, but no luck. I'm not sure if this is an issue on your machine, but it also seems to struggle with canceling the execution of the tests. Have you be able to run tests isolation to see which is causing the problem?

**Comment by eltherion** [comment]
@joules-o I managed to narrow tests compilation issue in scala 2.13 down to 2 tests. If you comment out tests below that line: https://github.com/krzemin/octopus/blob/8aa79b315f7e810f896ca97348b17b34332a9491/octopus/src/test/scala/octopus/ValidationSpec.scala#L373 it should compile (Ofc, when you get rid off `.right` as well. You can replace it by `.fold`). 

@krzemin Maybe you could help at this point, it seems that successful compilation for Scala 2.13 is really near.

**Comment by joules-o** [comment]
@eltherion Great! After commenting out those tests, there was one more issue in `AsyncValidationSpec` on line 32: 
```
method groupBy in class StringOps is deprecated (since 2.13.0): 

Use `s.toSeq.groupBy(...).view.mapValues(_.unwrap)` instead of `s.groupBy(...)`
```

I found that replacing `.groupBy(identity)` with either
`.toSeq.groupBy(identity)` or `.toSeq.groupBy(identity).view.mapValues(_.unwrap)`, means all tests pass!

**Comment by eltherion** [comment]
That's correct, but leaving tests for big case classes commented out is certainly not something that we would like to expect nor preserve (at least in case of the author). It has to be somehow resolved, unfortunately I have no idea how.

**Comment by joules-o** [comment]
I'm not sure either. I looked through the 2.13 changelogs for changes relating to big case classes but didn't see anything. Could it be something about the changes relating to macros, or collections?

**Comment by krzemin** [comment]
Fixed by #30.
